plugin/src/main/kotlin/org/jlleitschuh/gradle/ktlint/worker/KtLintWorkAction.kt

```diff
@@ -46,35 +37,74 @@ abstract class KtLintWorkAction : WorkAction<KtLintWorkAction.KtLintWorkParamete
+                ktlintInvoker.initialize(
+                    editorConfigPath = additionalEditorConfig,
+                    ruleSets = loadRuleSetsFromClasspathWithRuleSetProvider().filterRules(
+                        parameters.enableExperimental.getOrElse(false),
+                        parameters.disabledRules.getOrElse(emptySet())
+                    ),
+                    userData = userData,
+                    debug = debug
+                )
+            }
+
+            is ExperimentalParamsProviderInvocation -> {
+                ktlintInvoker.initialize(
+                    editorConfigPath = additionalEditorConfig,
+                    ruleProviders = loadRuleSetsFromClasspathWithRuleSetProviderV2().filterRules(
+                        parameters.enableExperimental.getOrElse(false),
+                        parameters.disabledRules.getOrElse(emptySet())
+                    ).flatten().toSet(),
+                    userData = userData,
+                    debug = debug
+                )
+            }
+
+            is RuleEngineInvocation -> {
+                ktlintInvoker.initialize(
+                    loadRuleSetsFromClasspathWithRuleSetProviderV2().filterRules(
+                        parameters.enableExperimental.getOrElse(false),
+                        parameters.disabledRules.getOrElse(emptySet())
+                    ).flatten().toSet(), userData
+                )
+            }
+        }
 
         parameters.filesToLint.files.forEach {
             val errors = mutableListOf<Pair<LintError, Boolean>>()
-            val ktLintParameters = KtLint.Params(
-                fileName = it.absolutePath,
-                text = it.readText(),
-                ruleSets = ruleSets,
-                userData = userData,
-                debug = debug,
-                editorConfigPath = additionalEditorConfig,
-                script = !it.name.endsWith(".kt", ignoreCase = true),
-                cb = { lintError, isCorrected ->
-                    errors.add(lintError to isCorrected)
-                }
-            )
 
             try {
                 if (formatSource) {
                     val currentFileContent = it.readText()
-                    val updatedFileContent = KtLint.format(ktLintParameters)
+                    val updatedFileContent = ktlintInvoker.invokeFormat(it) { lintError, isCorrected ->
+                        errors.add(lintError to isCorrected)
+                    }
 
                     if (updatedFileContent != currentFileContent) {
                         formattedFiles[it] = contentHash(it)
                         it.writeText(updatedFileContent)
                     }
                 } else {
-                    KtLint.lint(ktLintParameters)
+                    ktlintInvoker.invokeLint(it) { lintError, isCorrected ->
+                        errors.add(lintError to isCorrected)
+                    }
                 }
-            } catch (e: ParseException) {
```

Comment:
ParseException was changed to KtLintParseException, so this will catch either one